### PR TITLE
feat(cpu): add CPU support for draft model speculative decoding

### PR DESCRIPTION
**Pull request by ganeshr10**
## Purpose
This PR enables speculative decoding with draft models on CPU by adding PyTorch fallbacks for Triton kernels.

## Benchmark Test

Command:
``vllm serve Qwen/Qwen3-32B --dtype=bfloat16 --trust_remote_code --host 0.0.0.0 --port 8008  --max-model-len 20000  --speculative_config '{"model": "Qwen/Qwen3-1.7B", "num_speculative_tokens": 3, "method": "draft_model", "dtype": "bfloat16", "max-model-len": 20000}'``

``vllm bench serve --dataset-name hf --dataset-path philschmid/mt-bench --model Qwen/Qwen3-8B --host 0.0.0.0 --port 8008 --num-prompts 80 --max-concurrency (1/100)  --temperature 0.0 --top-p 1.0``

## Results - Qwen 3 metrics

- Draft Model : Qwen/Qwen3-1.7B
- Target Model : Qwen/Qwen3-32B
- Hardware Used: AMD EPYC 9654 96-Core Processor

image
image

#### Performance Highlights: 
- Equivalent Time-To-First-Token (TTFT) except in high batch sizes.
- Better Time-Per-Output-Token (TPOT) in SD as compared to Autoregressive decoding even at higher concurrencies.
- Higher output and total token throughput across all concurrency levels with speculative decoding.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/v1/spec_decode/utils.py [hunk 1]
```diff
@@ -1,7 +1,82 @@
...
+import torch
+
 from vllm.triton_utils import tl, triton
 
+# Check if Triton is actually available for kernel execution
+try:
+    import triton.language as triton_lang
+    HAS_TRITON = hasattr(triton, 'jit') and callable(triton.jit)
+except (ImportError, AttributeError):
+    HAS_TRITON = False
```
[the remark this answers is not in the document]
Comment (ganeshr10) [reply vllm/v1/spec_decode/utils.py]: Fixed

File: vllm/v1/sample/rejection_sampler.py [hunk 2]
```diff
@@ -623,20 +667,183 @@ def sample_recovered_tokens(
...
+                target_prob = target_probs[token_idx]
+                prob = torch.maximum(target_prob - draft_prob, torch.zeros_like(target_prob))
+            
+            # Gumbel-max trick: argmax(prob / q)
+            q_vals = q[req_idx]
+            # Avoid division by zero: set q_vals == 0 to a small value
+            q_vals_safe = torch.where(q_vals > 0, q_vals, torch.tensor(1e-10, device=q_vals.device))
+            recovered_id = torch.argmax(prob / q_vals_safe).item()
+            output_token_ids[token_idx] = recovered_id
+
```
[the remark this answers is not in the document]
Comment (ganeshr10) [reply vllm/v1/sample/rejection_sampler.py]: Done.

**Review by benchislett — CHANGES_REQUESTED** [review]
I am in the process of reworking the way we do `merge_toks_kernel` into a new kernel that can handle parallel drafting. Stay tuned for updates, but in the meantime (few days at most) there's not much value in maintaining this.

File: vllm/v1/sample/rejection_sampler.py [hunk 3]
```diff
@@ -19,6 +19,24 @@
...
+try:
+    import importlib.util
+
+    HAS_TRITON = (
+        importlib.util.find_spec("triton.language") is not None
+        and hasattr(triton, "jit")
+        and callable(triton.jit)
+    )
+except (ImportError, AttributeError):
+    HAS_TRITON = False
```
Comment (xuechendi) [inline vllm/v1/sample/rejection_sampler.py]: This is already a done func at https://github.com/vllm-project/vllm/blob/main/vllm/triton_utils/importing.py#L12
Is it ok to use the exsiting one in importing.py?

File: vllm/v1/sample/rejection_sampler.py [hunk 4]
```diff
@@ -385,15 +403,25 @@ def rejection_sample(
...
-        rejection_greedy_sample_kernel[(batch_size,)](
-            output_token_ids,
-            cu_num_draft_tokens,
-            draft_token_ids,
-            target_argmax,
-            bonus_token_ids,
-            is_greedy,
-            max_spec_len,
-        )
+        if HAS_TRITON and device.type == "cuda":
```
Comment (xuechendi) [inline vllm/v1/sample/rejection_sampler.py]:
```suggestion
        if HAS_TRITON:
```

Comment (xuechendi) [reply vllm/v1/sample/rejection_sampler.py]: Please remove `device.type == "cuda":`

File: vllm/v1/sample/rejection_sampler.py [hunk 5]
```diff
@@ -424,20 +452,34 @@ def rejection_sample(
...
-        target_probs,
-        bonus_token_ids,
-        recovered_token_ids,
-        uniform_probs,
-        is_greedy,
-        max_spec_len,
-        vocab_size,
-        NO_DRAFT_PROBS=draft_probs is None,
-    )
+    if HAS_TRITON and device.type == "cuda":
```
Comment (xuechendi) [inline vllm/v1/sample/rejection_sampler.py]:
```suggestion
    if HAS_TRITON:
```

File: vllm/v1/sample/rejection_sampler.py [hunk 6]
```diff
@@ -528,14 +570,23 @@ def expand_batch_to_tokens(
...
     expanded_x = x.new_empty(num_tokens)
-    expand_kernel[(batch_size,)](
-        expanded_x,
-        x,
-        cu_num_tokens,
-        replace_from,
-        replace_to,
-        MAX_NUM_TOKENS=MAX_SPEC_LEN,  # To avoid recompilation.
-    )
+    if HAS_TRITON and x.device.type == "cuda":
```
Comment (xuechendi) [inline vllm/v1/sample/rejection_sampler.py]:
```suggestion
    if HAS_TRITON:
```

File: vllm/v1/sample/rejection_sampler.py [hunk 7]
```diff
@@ -624,20 +675,196 @@ def sample_recovered_tokens(
...
-        cu_num_draft_tokens,
-        draft_token_ids,
-        draft_probs,
-        target_probs,
-        q,
-        vocab_size,
-        triton.next_power_of_2(vocab_size),
-        NO_DRAFT_PROBS=draft_probs is None,
-    )
+    if HAS_TRITON and device.type == "cuda":
```
Comment (xuechendi) [inline vllm/v1/sample/rejection_sampler.py]:
```suggestion
    if HAS_TRITON:
```

File: vllm/v1/sample/rejection_sampler.py [hunk 8]
```diff
@@ -624,20 +675,196 @@ def sample_recovered_tokens(
...
+        if is_greedy is not None and is_greedy[req_idx]:
+            continue
+
+        # Get start and end indices for this request
+        start_idx = 0 if req_idx == 0 else cu_num_draft_tokens[req_idx - 1].item()
+        end_idx = cu_num_draft_tokens[req_idx].item()
+        num_draft_tokens = end_idx - start_idx
+
+        rejected = False
+        for pos in range(num_draft_tokens):
```
Comment (xuechendi) [inline vllm/v1/sample/rejection_sampler.py]: same concern here

File: vllm/v1/sample/rejection_sampler.py [hunk 9]
```diff
@@ -624,20 +675,196 @@ def sample_recovered_tokens(
...
+            if not rejected:
+                draft_token_id = draft_token_ids[start_idx + pos].item()
+                target_argmax_id = target_argmax[start_idx + pos].item()
+                output_token_ids[req_idx, pos] = target_argmax_id
+                if draft_token_id != target_argmax_id:
+                    rejected = True
```
Comment (xuechendi) [inline vllm/v1/sample/rejection_sampler.py]: Is for loop a good impl here? Do we consider this func to be used by other devices except for CPU?
Wondering if masking on tensor might be more efficient? (Less CPU calls and possible to run on any device)

File: vllm/v1/spec_decode/eagle.py [hunk 10]
```diff
@@ -775,30 +777,51 @@ def set_inputs_first_pass(
...
-                parallel_drafting_token_id=self.parallel_drafting_token_id,
-                # Sizing info
-                # Note that we can deduce batch_size for free from the grid size
-                total_input_tokens=total_num_input_tokens,
-                num_padding_slots_per_request=self.extra_slots_per_request,
-                shift_input_ids=self.pass_hidden_states_to_model,
-                BLOCK_SIZE_TOKENS=BLOCK_SIZE_TOKENS,
-            )
+
+            if HAS_TRITON and self.device.type == "cuda":
```
Comment (xuechendi) [inline vllm/v1/spec_decode/eagle.py]:
```suggestion
            if HAS_TRITON:
```

File: vllm/v1/spec_decode/eagle.py [hunk 11]
```diff
@@ -917,20 +940,32 @@ def prepare_next_token_ids_padded(
...
-            backup_tokens_gpu,
-            next_token_ids,
-            valid_sampled_tokens_count,
-            gpu_input_batch.vocab_size,
-            num_tokens,
-            batch_size,
-            sampled_token_ids.stride(0),
-            BLOCK_SIZE_TOKENS=BLOCK_SIZE_TOKENS,
-        )
+        if HAS_TRITON and device.type == "cuda":
```
Comment (xuechendi) [inline vllm/v1/spec_decode/eagle.py]:
```suggestion
        if HAS_TRITON:
```

File: vllm/v1/spec_decode/eagle.py [hunk 12]
```diff
@@ -959,14 +994,24 @@ def prepare_inputs_padded(
...
         grid = (num_reqs,)
-        eagle_prepare_inputs_padded_kernel[grid](
-            spec_decode_metadata.cu_num_draft_tokens,
-            valid_sampled_tokens_count,
-            common_attn_metadata.query_start_loc,
-            token_indices_to_sample,
-            num_rejected_tokens_gpu,
-            num_reqs,
-        )
+        if HAS_TRITON and device.type == "cuda":
```
Comment (xuechendi) [inline vllm/v1/spec_decode/eagle.py]:
```suggestion
        if HAS_TRITON:
```

File: vllm/v1/spec_decode/utils.py [hunk 13]
```diff
@@ -3,13 +3,222 @@
...
+try:
+    import importlib.util
+
+    HAS_TRITON = (
+        importlib.util.find_spec("triton.language") is not None
+        and hasattr(triton, "jit")
+        and callable(triton.jit)
+    )
+except (ImportError, AttributeError):
+    HAS_TRITON = False
```
Comment (xuechendi) [inline vllm/v1/spec_decode/utils.py]: Is it ok to use the exsiting one - "HAS_TRITON" in importing.py?

File: vllm/v1/spec_decode/utils.py [hunk 14]
```diff
@@ -3,13 +3,222 @@
...
+def eagle_prepare_inputs_padded_pytorch(
+    cu_num_draft_tokens: torch.Tensor,  # [num_reqs]
+    valid_sampled_tokens_count: torch.Tensor,  # [num_reqs]
+    query_start_loc_gpu: torch.Tensor,  # [num_reqs + 1]
+    token_indices_to_sample: torch.Tensor,  # [num_reqs] (output)
+    num_rejected_tokens_gpu: torch.Tensor,  # [num_reqs] (output)
+    num_reqs: int,
+):
+    """PyTorch implementation of eagle_prepare_inputs_padded_kernel."""
+    for req_idx in range(num_reqs):
```
Comment (xuechendi) [inline vllm/v1/spec_decode/utils.py]: possible to use tensor masking instead of for loop?

File: vllm/v1/spec_decode/utils.py [hunk 15]
```diff
@@ -3,13 +3,222 @@
...
+    discard_request_mask: torch.Tensor,  # [num_reqs]
+    backup_next_token_ids: torch.Tensor,  # [num_reqs]
+    next_token_ids: torch.Tensor,  # [num_reqs] (output)
+    valid_sampled_tokens_count: torch.Tensor,  # [num_reqs] (output)
+    vocab_size: int,
+    num_sampled_tokens_per_req: int,
+    num_reqs: int,
+):
+    """PyTorch implementation of eagle_prepare_next_token_padded_kernel."""
+    for req_idx in range(num_reqs):
```
Comment (xuechendi) [inline vllm/v1/spec_decode/utils.py]: possible to use tensor masking instead of for loop?

File: vllm/v1/sample/rejection_sampler.py (hunk 3 above)
Comment (ganeshr10) [reply vllm/v1/sample/rejection_sampler.py]: I have updated the code by using the existing one.

File: vllm/v1/sample/rejection_sampler.py (hunk 4 above)
Comment (ganeshr10) [reply vllm/v1/sample/rejection_sampler.py]: Removed.

File: vllm/v1/sample/rejection_sampler.py (hunk 5 above)
Comment (ganeshr10) [reply vllm/v1/sample/rejection_sampler.py]: Removed.

File: vllm/v1/spec_decode/eagle.py (hunk 12 above)
Comment (ganeshr10) [reply vllm/v1/spec_decode/eagle.py]: Removed.

File: vllm/v1/spec_decode/eagle.py (hunk 11 above)
Comment (ganeshr10) [reply vllm/v1/spec_decode/eagle.py]: Removed.

File: vllm/v1/spec_decode/eagle.py (hunk 10 above)
Comment (ganeshr10) [reply vllm/v1/spec_decode/eagle.py]: Removed.

File: vllm/v1/sample/rejection_sampler.py (hunk 9 above)
Comment (ganeshr10) [reply vllm/v1/sample/rejection_sampler.py]: Thanks for the suggestion. I have updated the code to use tensor masking.

File: vllm/v1/sample/rejection_sampler.py (hunk 8 above)
Comment (ganeshr10) [reply vllm/v1/sample/rejection_sampler.py]: Thanks for the suggestion. I have updated the code to use tensor masking.

File: vllm/v1/sample/rejection_sampler.py (hunk 7 above)
Comment (ganeshr10) [reply vllm/v1/sample/rejection_sampler.py]: Removed.

File: vllm/v1/sample/rejection_sampler.py (hunk 6 above)
Comment (ganeshr10) [reply vllm/v1/sample/rejection_sampler.py]: Removed.

File: vllm/v1/spec_decode/utils.py (hunk 13 above)
Comment (ganeshr10) [reply vllm/v1/spec_decode/utils.py]: Updated the code to use the existing one.

File: vllm/v1/spec_decode/utils.py (hunk 14 above)
Comment (ganeshr10) [reply vllm/v1/spec_decode/utils.py]: Done. Updated the code to use tensor masking.

File: vllm/v1/spec_decode/utils.py (hunk 15 above)
Comment (ganeshr10) [reply vllm/v1/spec_decode/utils.py]: Done. Updated the code to use tensor masking.

**Review by benchislett — DISMISSED** [review]
This PR effectively doubles the amount of code to maintain/test for the speculative decoding inputs kernels. I am not sure this is worth the effort right now, especially as triton kernels are becoming the default way of handling inputs (see ModelRunnerV2) and maintaining separate pytorch implementations for them all can get burdensome

**Comment by bigPYJ1151** [comment]
Perhaps abstract kernels to some unified APIs and integrate torch ops via monkey patch will help to reduce the invasive and reuse kernel unrelated codes.

File: vllm/v1/spec_decode/eagle.py [hunk 16]
```diff
@@ -717,14 +715,15 @@ def set_inputs_first_pass(
...
             copy_and_expand_eagle_inputs_kernel[grid](
                 # (Padded) Inputs from the target model
                 target_token_ids_ptr=target_token_ids,
                 target_positions_ptr=target_positions,
-                next_token_ids_ptr=next_token_ids,  # sampled tokens, one per request
+                next_token_ids_ptr=next_token_ids,
                 # Outputs to the drafting buffers
                 out_input_ids_ptr=self.input_ids,
-                out_positions_ptr=self.positions,  # Doesn't support mrope for now
+                out_positions_ptr=self.positions,
```
Comment (bigPYJ1151) [inline vllm/v1/spec_decode/eagle.py]: Please keep these comments.

File: vllm/v1/spec_decode/eagle.py [hunk 17]
```diff
@@ -735,7 +734,6 @@ def set_inputs_first_pass(
                 padding_token_id=0,
                 parallel_drafting_token_id=self.parallel_drafting_token_id,
                 # Sizing info
-                # Note that we can deduce batch_size for free from the grid size
```
Comment (bigPYJ1151) [inline vllm/v1/spec_decode/eagle.py:738]: Same as above.

File: vllm/v1/worker/cpu_model_runner.py [hunk 18]
```diff
@@ -23,7 +25,7 @@ def __init__(self, vllm_config: VllmConfig, device: torch.device):
             super().__init__(vllm_config, device)
 
         assert device == torch.device("cpu")
-        assert self.speculative_config is None, "spec decode is not supported."
+        # Note: speculative decoding is now supported on CPU with PyTorch fallbacks
```
Comment (bigPYJ1151) [inline vllm/v1/worker/cpu_model_runner.py]: You have converted them to C++ native impls

File: vllm/v1/worker/cpu_model_runner.py [hunk 19]
```diff
@@ -102,6 +161,75 @@ def _zero_block_ids(self, block_ids: list[int]) -> None:
         # so stale KV cache data never affects computation.
         pass
 
+    def get_dp_padding(self, num_tokens: int) -> tuple[int, torch.Tensor | None]:
+        # Note: For CPU backend, dp padding is not required for now.
+        return 0, None
```
Comment (bigPYJ1151) [inline vllm/v1/worker/cpu_model_runner.py]: It has been removed in #38383

File: csrc/cpu/spec_decode_utils.cpp [hunk 20]
```diff
@@ -0,0 +1,447 @@
+#include <torch/extension.h>
+#include <omp.h>
```
Comment (bigPYJ1151) [inline csrc/cpu/spec_decode_utils.cpp]: Some compilation errors in the build image with the headers. Please use ```cpu_types.hpp```, which already includes required headrs.

**Review by bigPYJ1151 — COMMENTED** [review]
Thanks @ganeshr10 ! It looks good now. Just have some nits, please check :)

File: vllm/v1/worker/cpu_model_runner.py (hunk 19 above)
Comment (ganeshr10) [reply vllm/v1/worker/cpu_model_runner.py]: Removed in my code as well.

File: vllm/v1/worker/cpu_model_runner.py (hunk 18 above)
Comment (ganeshr10) [reply vllm/v1/worker/cpu_model_runner.py]: Fixed.

File: vllm/v1/spec_decode/eagle.py (hunk 16 above)
Comment (ganeshr10) [reply vllm/v1/spec_decode/eagle.py]: Done. Retained the comments

File: vllm/v1/spec_decode/eagle.py (hunk 17 above)
Comment (ganeshr10) [reply vllm/v1/spec_decode/eagle.py:738]: Done.

**Review by bigPYJ1151 — APPROVED** [review]

**Comment by bigPYJ1151** [comment]
Hi @benchislett I think the new implementation has resolved your concern. I would move the PR forward. Please let me know if you have further thoughts, thanks! :)
